node/pkg/celo/by_transaction.go

```diff
@@ -0,0 +1,98 @@
+package celo
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	celo_common "github.com/celo-org/celo-blockchain/common"
+	"github.com/celo-org/celo-blockchain/ethclient"
+	"github.com/certusone/wormhole/node/pkg/celo/abi"
+	"github.com/certusone/wormhole/node/pkg/common"
+	"github.com/certusone/wormhole/node/pkg/vaa"
+	eth_common "github.com/ethereum/go-ethereum/common"
+)
+
+var (
+	// SECURITY: Hardcoded ABI identifier for the LogMessagePublished topic. When using the watcher, we don't need this
+	// since the node will only hand us pre-filtered events. In this case, we need to manually verify it
+	// since ParseLogMessagePublished will only verify whether it parses.
+	logMessagePublishedTopic = celo_common.HexToHash("0x6eb224fb001ed210e379b335e35efe88672a8ce935d981a6896b27ffdf52a3b2")
+)
+
+// MessageEventsForTransaction returns the lockup events for a given transaction.
+// Returns the block number and a list of MessagePublication events.
+func MessageEventsForTransaction(
+	ctx context.Context,
+	c *ethclient.Client,
+	contract celo_common.Address,
+	chainId vaa.ChainID,
+	tx celo_common.Hash) (uint64, []*common.MessagePublication, error) {
+
+	f, err := abi.NewAbiFilterer(contract, c)
+	if err != nil {
+		return 0, nil, fmt.Errorf("failed to create ABI filterer: %w", err)
+	}
+
+	// Get transactions logs from transaction
+	receipt, err := c.TransactionReceipt(ctx, tx)
+	if err != nil {
+		return 0, nil, fmt.Errorf("failed to get transaction receipt: %w", err)
+	}
+
+	// Bail early when the transaction receipt status is anything other than
+	// 1 (success). In theory, this check isn't strictly necessary - a failed
+	// transaction cannot emit logs and will trigger neither subscription
+	// messages nor have log messages in its receipt.
+	//
+	// However, relying on that invariant is brittle - we connect to a lot of
+	// EVM-compatible chains which might accidentally break this API contract
+	// and return logs for failed transactions. Check explicitly instead.
+	if receipt.Status != 1 {
+		return 0, nil, fmt.Errorf("non-success transaction status: %d", receipt.Status)
+	}
+
+	// Get block
+	block, err := c.BlockByHash(ctx, receipt.BlockHash)
+	if err != nil {
+		return 0, nil, fmt.Errorf("failed to get block: %w", err)
+	}
+
+	msgs := make([]*common.MessagePublication, 0, len(receipt.Logs))
+
+	// Extract logs
+	for _, l := range receipt.Logs {
+		// SECURITY: Skip logs not produced by our contract.
+		if l.Address != contract {
+			continue
+		}
```

Comment:
Also @claudijd ^

We probably want tests for these security bits. Tagging for later.

Reply:
yeah, eth/by_transaction works the same.  Was chatting about this with Leo this AM, sounds like we might benefit from a refactor on this to make it more testable.  However, I don't think it should block this PR.